### PR TITLE
Implementation of Git middleware

### PR DESCRIPTION
**Pull request by abiosoft**
Caddyfile Syntax :
```
git repo path {
	repo
	path
	branch
	key
	interval
}
```
`repo` 	- git repository
	compulsory. Both ssh (e.g. git@github.com:user/project.git)
	and https(e.g. https://github.com/user/project) are supported.
	Can be specified in either config block or top level

`path` 	- directory to pull into
	optional.  Defaults to site root.

`branch` 	- git branch or tag
	optional. Defaults to master

`key` 	- path to private ssh key
	optional. Required for private repositories. e.g. /home/user/.ssh/id_rsa

`interval`  - interval between git pulls in seconds
	optional. Defaults to 3600 (1 Hour).

Examples :

public repo pulled into site root
```
git github.com/user/myproject
```
public repo pulled into mysite
```
git https://github.com/user/myproject mysite
```
private repo pulled into mysite with tag v1.0 and interval of 1 day
```
git {
	repo 	git@github.com:user/myproject
	branch 	v1.0
	path	mysite
	key 	/home/user/.ssh/id_rsa
	interval 86400 # 1 day
}
```
Caddyfile with private git repo and php support via fastcgi.
path defaults to /var/www/html/myphpsite as specified in root config.
```
0.0.0.0:8080

git {
	repo 	git@github.com:user/myphpsite
	key 	/home/user/.ssh/id_rsa
	interval 86400 # 1 day
}

fastcgi / 127.0.0.1:9000 php

root /var/www/html/myphpsite
```
A pull is first attempted after initialization. Afterwards, a pull is attempted
after request to server and if time taken since last successful pull is higher than interval.

After the first successful pull (should be during initialisation except an error occurs), subsequent pulls are done in background and do not impact request time.

Note: private repositories are currently only supported and tested on Linux and OSX

### CODE REVIEW

File: middleware/git/git.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+// Git represents a middleware instance that pulls git repository.
+type Git struct {
+	Next middleware.Handler
+	Repo *Repo
+}
+
+// ServeHTTP satisfies the middleware.Handler interface.
+func (g Git) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
+	if err := g.Repo.Pull(); err != nil {
+		log.Println(err)
```
Comment (mholt) [inline middleware/git/git.go]: Conveniently, this line is not needed. Just return the err like you do in the next line. Error handling middleware will take care of logging it.

File: config/middleware.go [hunk 2]
```diff
@@ -48,6 +49,7 @@ func init() {
 	register("ext", extensions.New)
 	register("basicauth", basicauth.New)
 	register("proxy", proxy.New)
+	register("git", git.New)
```
Comment (mholt) [inline config/middleware.go]: Good. I think this is a good place in the chain for this middleware to be registered.

File: middleware/git/git.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			case "repo":
+				if !c.NextArg() {
+					return nil, c.ArgErr()
+				}
+				repo.Url = c.Val()
+			case "path":
+				if !c.NextArg() {
+					return nil, c.ArgErr()
+				}
+				repo.Path = c.Val()
```
Comment (mholt) [inline middleware/git/git.go]: What do you think about making the path relative to site root, since it specifies where in the site to clone the repo?

Comment (abiosoft) [inline middleware/git/git.go]: I think it makes more sense that way.

**Pull request merged**